src/main/java/io/cryostat/discovery/KubeApiDiscovery.java

```diff
@@ -0,0 +1,623 @@
+import org.apache.commons.lang3.tuple.Triple;
+import org.eclipse.microprofile.config.inject.ConfigProperty;
+import org.jboss.logging.Logger;
+
+@ApplicationScoped
+public class KubeApiDiscovery {
+    public static final String REALM = "KubernetesApi";
+
+    public static final long ENDPOINTS_INFORMER_RESYNC_PERIOD = Duration.ofSeconds(30).toMillis();
+
+    public static final String DISCOVERY_NAMESPACE_LABEL_KEY = "discovery.cryostat.io/namespace";
+
+    @Inject Logger logger;
+
+    @Inject KubeConfig kubeConfig;
+
+    @ConfigProperty(name = "cryostat.discovery.kubernetes.enabled")
+    boolean enabled;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-names")
+    Optional<List<String>> JmxPortNames;
+
+    @ConfigProperty(name = "cryostat.discovery.k8s.port-numbers")
+    Optional<List<Integer>> JmxPortNumbers;
+
+    private final Map<Triple<String, String, String>, Pair<HasMetadata, DiscoveryNode>>
+            discoveryNodeCache = new ConcurrentHashMap<>();
+    private final Map<Triple<String, String, String>, Object> queryLocks =
+            new ConcurrentHashMap<>();
+
+    private final LazyInitializer<HashMap<String, SharedIndexInformer<Endpoints>>> nsInformers =
+            new LazyInitializer<HashMap<String, SharedIndexInformer<Endpoints>>>() {
+                @Override
+                protected HashMap<String, SharedIndexInformer<Endpoints>> initialize()
+                        throws ConcurrentException {
+                    // TODO: add support for some wildcard indicating a single Informer for any
+                    // namespace that Cryostat has permissions to. This will need some restructuring
+                    // of how the namespaces within the discovery tree are mapped.
+                    var result = new HashMap<String, SharedIndexInformer<Endpoints>>();
+                    kubeConfig
+                            .getWatchNamespaces()
+                            .forEach(
+                                    ns -> {
+                                        result.put(
+                                                ns,
+                                                client().endpoints()
+                                                        .inNamespace(ns)
+                                                        .inform(
+                                                                new EndpointsHandler(),
+                                                                ENDPOINTS_INFORMER_RESYNC_PERIOD));
+                                        logger.infov(
+                                                "Started Endpoints SharedInformer for"
+                                                        + " namespace \"{0}\"",
+                                                ns);
+                                    });
+                    return result;
+                }
+            };
+
+    @Transactional
+    void onStart(@Observes @Priority(1) StartupEvent evt) {
```

Comment:
What is this Priority for?

Reply:
Oh this is to ensure the Realm node is persisted first before initializing the informers. During my testing, I see that the informers call the event handler right away when created so if it's called in the same transaction as the realm node creation, the realm is not yet available.

Reply:
maybe there is a nicer way around it?

Reply:
Okay, makes sense. Just add a comment here explaining why it has the Priority then please.